test/test_type_promotion.py

```diff
@@ -430,6 +431,123 @@ def test_transpose(self, device):
         a = torch.tensor([[True, True], [False, True]], device=device)
         self.assertEqual(a.t() == 0, a.t() == False)  # noqa: E712
 
+    def _test_sparse_op_input_tensors(self, device, dtype, coalesced, zeros=True):
+        t = self._get_test_tensor(device, dtype, not zeros)
+        if zeros and dtype != torch.bool:
+            # ensure sparsity. Bool should already have sufficient sparsity.
+            mask = self._get_test_tensor(device, torch.bool)
+            t = t * mask
+        if coalesced:
+            s = t.to_sparse()
+        else:
+            s = t.to_sparse()
+            indices = torch.cat((s.indices(), s.indices()), 1)
+            values = torch.cat((s.values(), s.values()), 0)
+            s = torch.sparse_coo_tensor(indices=indices, values=values, size=s.size(), dtype=dtype, device=device)
+            t = s.to_dense()
+        self.assertEqual(s.is_coalesced(), coalesced)
+        self.assertEqual(s.dtype, dtype)
+        self.assertEqual(t.dtype, s.dtype)
+        return (t, s)
+
+    def _get_precision(self, dtype, coalesced):
+        if dtype == torch.half and not coalesced:
+            # very low precision for uncoalesced float16 sparse tensors since
+            # ops like (s1 + s2).to_dense() will add four low-precision
+            # floating point values.
+            return 5e-2
+        if dtype == torch.half:
+            return 1e-3
+        return 1e-5
+
+    def _test_sparse_op(self, op_name, inplace, dtype1, dtype2, device, coalesced):
```

Comment:
I still think the special casing in this function makes it hard to read. We should optimize for readability of the tests so that future readers don't have to go through all of the lines here to figure out what combination of (op, input1_sparsity, input2_sparsity) is supported.

If we had something like a table that showed exactly what we support:
```
tests = {
    ('add', [SPARSE, SPARSE], NO_ERROR, ALL_DTYPES)
    ('sub', [SPARSE, SPARSE], NO_ERROR, ALL_DTYPES)
    ('mul', [SPARSE, SPARSE], NO_ERROR, ALL_DTYPES)
    ('div', [SPARSE, SPARSE], UNSUPPORTED, ALL_DTYPES)

    ('add', [DENSE, SPARSE], NO_ERROR, ALL_DTYPES)
    ('sub', [DENSE, SPARSE], NO_ERROR, ALL_DTYPES)
    ('mul', [DENSE, SPARSE], UNSUPPORTED, ALL_DTYPES)
    ('div', [DENSE, SPARSE], UNSUPPORTED, ALL_DTYPES)

    ('add', [SPARSE, DENSE], UNSUPPORTED, ALL_DTYPES)
    ('sub', [SPARSE, DENSE], UNSUPPORTED, ALL_DTYPES)
    ('mul', [SPARSE, DENSE], UNSUPPORTED, ALL_DTYPES)
    ('div', [SPARSE, DENSE_NOZEROS], UNSUPPORTED, ALL_DTYPES)

    ('add', [SPARSE, SCALAR], UNSUPPORTED, ALL_DTYPES)
    ('sub', [SPARSE, SCALAR], UNSUPPORTED, ALL_DTYPES)
    ('mul', [SPARSE, SCALAR], NO_ERROR, EXCEPT_HALF)
    ('div', [SPARSE, SCALAR_NOZERO], NO_ERROR, EXCEPT_HALF)
}
```
then we could run each of those through a function `run_test(op, inputs, expected_behavior, dtypes)` that doesn't have to do special casing.

In order to construct this table, I spent 10 minutes reading through ~70 LOC in this test to understand what is supported or not. From this table, a future code reader will only have to read ~16 lines and understand in less than one minute what is supported. (If you asked me "Do we support sparse-dense multiplication", it's easy to figure that out with the table).

I proposed this before in https://github.com/pytorch/pytorch/pull/30317#discussion_r350339659 and you had some concerns about this approach that I'd like to address here. 

>  My main concern with this approach though would be having the list of tests hard-coded, since it leaves a lot of room for missing a case. 

I agree it is more difficult to write tests correctly in this style but we should be optimizing for readability. We write code once, but future contributors will read the code more than once. Code review can also help against missing cases.

> There's a fair bit that would also still need to be added too (e.g. can_cast errors in the next diff are dependent on the dtypes of the operands), so I think this would end up being just as much code. 

The main priority is readability, not LOC.

> Plus by being specific the comments can describe each particular case in the right place

To me, the table looks self documenting. If we want to put extra comments (I noticed that your code even has error messages, which is great!) we could just add them as comments to the table or even add the error messages into the table.

> and by re-using test tensors created for out of place operations, we save a bit of run-time over creating new test tensors for each test.

We can still generate inputs once in this approach. For example, `SPARSE` can just be a random sparse tensor that is generated once rather than an enum / string. 

What do you think? I'm open to more discussion and feedback. I'd really like to see less special-casing like this in our testing codebase.

Reply:
> I agree it is more difficult to write tests correctly in this style

A big part of how I think about readability is the ease of maintaining correctness in the future, not just about the ability to consume the test-coverage passively (though being able to confirm full-coverage at a glance also has value). If I were to write the test in the suggested style, I'd want to add a test to test-the-test, i.e. to confirm that the static test config contained all possible combinations. When I think what that would look like, a test validating that a static config contains what it says it contains, it doesn't look very nice to me.

I could imagine instead structuring the table as a map where we look up the metadata about how/if a particular combination should be tested, and missing-metadata fails the test. The test would still get called in a loop over all possible options, enforcing that no combination is missing.

Even with a test-of-the-test or a dictionary test-metadata-lookup structure, this would introduce some gaps. E.g. using what I assume is a list of dtypes for `EXCEPT_HALF`, we'd presumably be dropping coverage for half on cuda. If instead it's an enum, the special casing code doesn't go away.

> The main priority is readability, not LOC.

Sure, but you seem to be saying that special-cases impairs readability. I'm saying the special casing code doesn't go away. `if inplace and not torch.can_cast(common_dtype, dtype1):` doesn't disappear just by abstracting away the list of supported ops. Some special cases are inherent to fully covering all possibilities.

Even the main thing that would be abstracted away, whether an op is supported, seems like it could be clearer to me as `# op(sparse,sparse)\n if op_name != 'div':`, specifying the exception vs scanning through a full list of possible combinations. The table approach could also become harder to read if we do end up deciding we want to add mixed-type sparse coverage to other ops and include them in this test. The test as-written could also get a lot simpler if we decide to implement all/most of the unsupported cases.

> then we could run each of those through a function run_test(op, inputs, expected_behavior, dtypes) that doesn't have to do special casing.

It abstracts some of the special casing, but I don't think it removes all of it.

> SPARSE can just be a random sparse tensor that is generated once rather than an enum / string.

both sparse and dense tensors still needs to be re-generated/cloned for in-place tests. We could add special handling in the tests to clone only for in-place, but these are the kinds of conditions that make re-writing this not-necessarily as simple/clear as you're hoping.

If the tests and code are correct and properly cover all possible cases, I'd argue this is more of a subjective stylistic difference we could iterate on and think about, not necessarily something so obfuscated that it should block the PR (though I'll probably still change the in-place coalesce on `t` for this PR).

I think the style you're favoring may be partly informed by what has worked reasonably well for `method_tests()` in common_method_invocations, though I'm not entirely sure that approach is ideal even there. If we do land, maintain, and iterate on this test a bit, we might get to an approach we prefer or that changes our thinking on other test coverage. We can always change this to the tabular common-method-invocations style later if it's really not working.

Reply:
> A big part of how I think about readability is the ease of maintaining correctness in the future, not just about the ability to consume the test-coverage passively (though being able to confirm full-coverage at a glance also has value). If I were to write the test in the suggested style, I'd want to add a test to test-the-test, i.e. to confirm that the static test config contained all possible combinations. 

We don't need a test to validate the static config. How do you see it changing in the future? There are only this many cases so we won't add things to it, only modify it or simplify it as we improve unsupported operations. Because of that a test to validate the config is overkill.

> Even with a test-of-the-test or a dictionary test-metadata-lookup structure, this would introduce some gaps. E.g. using what I assume is a list of dtypes for EXCEPT_HALF, we'd presumably be dropping coverage for half on cuda. If instead it's an enum, the special casing code doesn't go away.

Sure, we could add another column for 'CUDA', and double the size of the table. I missed that we're testing different dtypes for CUDA, which is another argument against the readability of the current tests. There's always more information that we can add to the table to avoid special casing. 

> Sure, but you seem to be saying that special-cases impairs readability. I'm saying the special casing code doesn't go away. if inplace and not torch.can_cast(common_dtype, dtype1): doesn't disappear just by abstracting away the list of supported ops. Some special cases are inherent to fully covering all possibilities.

I think you're misunderstanding my definition of "special casing". What I would like to remove is different behavior in the test for each operator.  I wouldn't consider ` if inplace and not torch.can_cast(common_dtype, dtype1): ` to be special casing: that is common to all of the tests, no matter the operator. 

Examples of things that aren't common to each operator: the type of input (division by zero is not allowed), what dypes are supported on which devices (this varies between operators).

>  The test as-written could also get a lot simpler if we decide to implement all/most of the unsupported cases.

You've pointed out the primary reason why there is special casing: to support testing for unsupported cases. Let me rephrase why I wanted the table approach: it makes it easier for future maintainers to see exactly what the unsupported cases are.

> If the tests and code are correct and properly cover all possible cases, I'd argue this is more of a subjective stylistic difference we could iterate on and think about, not necessarily something so obfuscated that it should block the PR (though I'll probably still change the in-place coalesce on t for this PR).

I'm fine with not blocking this PR on this point. However, I'm not done reading through the PR and I don't want to open too many threads at once because then I can't keep track of all of them.

> I think the style you're favoring may be partly informed by what has worked reasonably well for method_tests() in common_method_invocations, though I'm not entirely sure that approach is ideal even there.

We have a few places in the codebase where we use tables instead of having special logic to handle unsupported operations: 
- https://github.com/pytorch/pytorch/blob/master/test/cpp_api_parity/parity-tracker.md
- https://github.com/pytorch/pytorch/blob/4fd20c0816d00e445d96085b9dc2fa7a443159ff/test/test_torch.py#L11423-L11484

Reply:
> How do you see it changing in the future?

All code is eventually modified. I could imagine us adding `pow`, comparison ops, `where`, any other binary op. I'm not sure what it will eventually become, this is just a start for mixed-type sparse support.

Reply:
I'd still like to do it in a follow-up (or maybe stacked) PR, but I think we might both being happy if I add a table including only the failure cases. That way the op-specific conditions are (mostly) removed from the main test, but we retain the constraint on coverage and don't have a giant list of working cases that could nearly be generated by a for loop. Newly added ops would have coverage enforced. Also, as we implement the cases that currently fail, the table would shrink.

Reply:
I'm fine with having the tests in their current state. Feel free to send a followup if you'd like, I'm curious how the table including failure cases will look like